File: src/parser/statements/file-statement.cpp
#include "parser/statements/file-statement.h"
#include "parser/statements/statement.h"
#include "transpiler/transpiler.h"
#include "util.h"

vector<Component*> FileStatement::nodes() const { 
    vector<Component*> nodes;
    nodes.reserve(statements.size());
    
    for(Node* node : statements) nodes.push_back(node);
    return nodes;
}

FileStatement::~FileStatement() {
    for(Component* ptr : statements) delete ptr;
}

Transpiler::Line FileStatement::transpile(Transpiler& transpiler) {
    for(Node* statement : statements) {
        transpiler.write(statement->transpile(transpiler).toString(), "");
    }

    return {};
}

void FileStatement::makeTree(Parser& parser) {
    while(not parser.next().has({""})) {
        Node* node = Statement::construct(parser);
        if(not node) return;
        statements.push_back(node);
    }
}

File: app/src/main/kotlin/sea/parser/statements/FileStatement.kt
package sea.grammar

import sea.parser.*
import sea.grammar.LineStatement

class FileStatement : Node() {
    val statements = ArrayList<Node>()

    override val parts: Parts
        get() = (statements.map { it })

    fun makeTree(parser: Parser) {
        while(!parser.next.has("EOF")) {
            statements.add(LineStatement.construct(parser))
        }
    }

    override fun publish(publisher: Publisher) {
        statements.forEach { it.publish(publisher) }
    }

    override fun transpile(transpiler: Transpiler): TExpression {
        statements.forEach { statement -> 
            transpiler.write(statement.transpile(transpiler).toString(), "")
        }

        return TExpression()
    }
}
